src/SignatureDB.sol

```diff
@@ -0,0 +1,105 @@
+        address gnosisSafe,
+        bytes32 dataHash,
+        bytes calldata data,
+        bytes calldata signatures
+    ) public {
+        checkSignaturesGnosis(gnosisSafe, dataHash, data, signatures);
+
+        // Once we have checked the signatures, we can store the initial transaction
+        Transaction memory transaction = Transaction(data, signatures);
+        transactionsForGnosisSafe[gnosisSafe][dataHash] = transaction;
+    }
+
+    /// @notice You can add one or multiple signatures via concatenated bytes
+    /// @dev The only requirement for `signatures` is that it is greater than
+    /// the previous bytes of signatures. This means that the set of signers
+    /// could change. So this does not strictly always add signatures. If the
+    /// new set of signatures is greater than the previous set of signatures but
+    /// has different signers, the signers from the previous set that are not
+    /// represented in the new set will be removed. What this does guarantee,
+    /// however, is that calling `addSignatures()` will always ensure that a
+    /// transaction is closer to execution than before.
+    /// TODO: Handle signature insertion on-chain, which ensure that signatures
+    /// are always increasing.
+    function addSignatures(
+        address gnosisSafe,
+        bytes32 dataHash,
+        bytes calldata signatures
+    ) public {
+        Transaction memory transaction = transactionsForGnosisSafe[gnosisSafe][
+            dataHash
+        ];
+
+        // Check whether the transaction has been initialized.
+        // We can't initialize the transaction for the user here because we
+        // don't have the data available. We avoid passing in the data to save
+        // on calldata costs, because the data is only necessary when a
+        // signature is initialized and then afterward is kept in storage
+        require(
+            transaction.signatures.length != 0,
+            "Transaction must be initialized"
+        );
+
+        // Place the new signatures at the start of the concatenated bytes
+        // This ensures that any failure in the new signatures will
+        // short-circuit. As a result, gas will not be wasted checking the old
+        // signatures.
+        // TODO: You'll need to deconstruct, sort, and insert these signatures
+        // on-chain to ensure that previous signatures are never removed
+        bytes memory signaturesConcat = bytes.concat(
+            signatures,
+            transaction.signatures
+        );
+
+        // Check that the new data is valid
+        // TODO: Confirm whether you can check only `signatures` for the new
+        // signatures, without needing to check all of `signaturesConcat`. You'd
+        // need to manually check whether the signatures meet the transaction
+        // threshold, which as a benefit also allows you to cut off signing when
+        // the transaction threshold is reached.
+        checkSignaturesGnosis(
```

Comment:
I don't think this will work since this contract will be on a different network than the Safe

Reply:
Haha same thinking! Had this in my notes simultaneously:
```
* checkNSignatures checks whether the Gnosis Safe exists on the same chain
* A key design constraint is that the Multisig Mempool should be able to work on different chains
* Instead, you can let the Safe itself sign a message specifying the owners. That can also then be relayed on-chain by anyone who wants to relay signatures 
```

Going to merge this PR so that I can then do another round of architecture updates that takes this into account.